subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/CompleteDirectorySnapshotTest.groovy

```diff
@@ -0,0 +1,180 @@
+
+        interaction {
+            invalidateDescendantOfSelectedChild(null)
+            noMoreInteractions()
+        }
+
+        where:
+        vfsSpec << filterSpecs(CHILD_IS_PREFIX)
+    }
+
+    def "storing for path #vfsSpec.absolutePath adds no information (#vfsSpec)"() {
+        setupTest(vfsSpec)
+
+        expect:
+        initialRoot.store(absolutePath, offset, CASE_SENSITIVE, Mock(MetadataSnapshot)) is initialRoot
+
+        where:
+        vfsSpec << filterSpecs(NO_COMMON_PREFIX + SAME_PATH + CHILD_IS_PREFIX)
+    }
+
+    def "querying the snapshot for non-existent child #vfsSpec.absolutePath yields a missing file (#vfsSpec)"() {
+        setupTest(vfsSpec)
+
+        when:
+        CompleteFileSystemLocationSnapshot foundSnapshot = initialRoot.getSnapshot(absolutePath, offset, CASE_SENSITIVE).get() as CompleteFileSystemLocationSnapshot
+        then:
+        foundSnapshot.type == FileType.Missing
+        foundSnapshot.absolutePath == absolutePath
+
+        where:
+        vfsSpec << filterSpecs(NO_COMMON_PREFIX)
+    }
+
+    def "querying the snapshot for child #vfsSpec.absolutePath yields the child (#vfsSpec)"() {
+        setupTest(vfsSpec)
+
+        when:
+        CompleteFileSystemLocationSnapshot foundSnapshot = initialRoot.getSnapshot(absolutePath, offset, CASE_SENSITIVE).get() as CompleteFileSystemLocationSnapshot
+        then:
+        foundSnapshot == selectedChild
+        1 * selectedChild.snapshot >> Optional.of(selectedChild)
+        interaction { noMoreInteractions() }
+
+        where:
+        vfsSpec << filterSpecs(SAME_PATH)
+    }
+
+    def "querying a snapshot in child #vfsSpec.absolutePath yields the found snapshot (#vfsSpec)"() {
+        setupTest(vfsSpec)
+        def grandChild = Mock(CompleteFileSystemLocationSnapshot)
+
+        when:
+        CompleteFileSystemLocationSnapshot foundSnapshot = initialRoot.getSnapshot(absolutePath, offset, CASE_SENSITIVE).get() as CompleteFileSystemLocationSnapshot
+        then:
+        foundSnapshot == grandChild
+        interaction {
+            getDescendantSnapshotOfSelectedChild(grandChild)
+            noMoreInteractions()
+        }
+
+        where:
+        vfsSpec << filterSpecs(CHILD_IS_PREFIX)
+    }
+
+    def "querying a non-existent snapshot in child #vfsSpec.absolutePath yields a missing snapshot (#vfsSpec)"() {
+        setupTest(vfsSpec)
+
+        when:
+        CompleteFileSystemLocationSnapshot foundSnapshot = initialRoot.getSnapshot(absolutePath, offset, CASE_SENSITIVE).get() as CompleteFileSystemLocationSnapshot
+        then:
+        foundSnapshot.type == FileType.Missing
+        foundSnapshot.absolutePath == absolutePath
+        interaction {
+            getDescendantSnapshotOfSelectedChild(null)
+            noMoreInteractions()
+        }
+
+        where:
+        vfsSpec << filterSpecs(CHILD_IS_PREFIX)
+    }
+
+    private static List<VirtualFileSystemTestSpec> filterSpecs(List<VirtualFileSystemTestSpec> fullList) {
```

Comment:
Can we capture in the name of this method as to how/why it filters the specs?

Reply:
Done.